staging/src/k8s.io/externaljwt/pkg/plugin/plugin.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"gopkg.in/square/go-jose.v2/jwt"
+	ejspb "k8s.io/externaljwt/apis/v1alpha1"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+type VerificationKeys struct {
+	Keys          []serviceaccount.PublicKey
+	DataTimestamp time.Time
+}
+
+type Metadata struct {
+	MaxTokenExpirationSeconds int
+}
+
+// New sets up all dependencies for the plugin and cache, in
+// order to minimize changes to upstream files.
+func New(ctx context.Context, issuer, socketPath string) (*Plugin, *KeyCache, error) {
+	conn, err := grpc.Dial(
+		socketPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(func(ctx context.Context, path string) (net.Conn, error) {
+			return (&net.Dialer{}).DialContext(ctx, "unix", path)
+		}),
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("while dialing grpc socket at %q: %w", socketPath, err)
+	}
+
+	plugin := newPlugin(issuer, conn)
+
+	initialFillCtx, cancel := context.WithTimeout(ctx, 60*time.Second)
+	defer cancel()
+
+	if err := plugin.keyCache.InitialFill(initialFillCtx); err != nil {
+		return nil, nil, fmt.Errorf("while initially filling key cache: %w", err)
+	}
+	go plugin.keyCache.StartPeriodicSync(ctx)
+
+	go func() {
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	return plugin, plugin.keyCache, nil
+}
+
+// enables pluging in an external jwt signer.
+type Plugin struct {
+	iss      string
+	client   ejspb.ExternalJWTSignerClient
+	keyCache *KeyCache
+}
+
+// newPlugin constructs an implementation of external JWT signer plugin.
+func newPlugin(iss string, conn *grpc.ClientConn) *Plugin {
+	client := ejspb.NewExternalJWTSignerClient(conn)
+	plugin := &Plugin{
+		iss:    iss,
+		client: client,
+	}
+	plugin.keyCache = newKeyCache(plugin)
+	return plugin
+}
+
+// GenerateToken creates a service account token with the provided claims by
+// calling out to the external signer binary.
+func (p *Plugin) GenerateToken(claims *jwt.Claims, privateClaims interface{}) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	claims.Issuer = p.iss
+
```

Comment:
can we unify the claim merging code between the file-based GenerateToken and here?

that would make the existing GenerateToken impl look like:

```go
mergedClaims, err := externaljwt.MergeClaims(privateClaims, claims, j.iss)
...
return jwt.Signed(p).Claims(mergedClaims).CompactSerialize()
```

and here would look like:
```go
mergedClaims, err := externaljwt.MergeClaims(privateClaims, claims, p.iss)
...
marshaledClaims, err := json.Marshal(mergedClaims)
...
encodedClaims, err := base64.RawURLEncoding.EncodeToString(mergedClaims)
...
... sign request
```

Reply:
I strongly prefer if we leave the existing merge code alone (which looks finicky), and instead force the external jwt code path to go through the same go-jose logic without us reimplementing any of it:

```go
// GenerateToken is shared between internal and external signer code.
func GenerateToken(signer jose.Signer, iss string, claims *jwt.Claims, privateClaims interface{}) (string, error) {
	// claims are applied in reverse precedence
	return jwt.Signed(signer).
		Claims(privateClaims).
		Claims(claims).
		Claims(&jwt.Claims{
			Issuer: iss,
		}).
		CompactSerialize()
}

func mergeClaims(iss string, claims *jwt.Claims, privateClaims interface{}) ([]byte, error) {
	var out []byte
	signer := payloadGrabber(func(payload []byte) { out = payload })
	_, err := GenerateToken(signer, iss, claims, privateClaims)
	if len(out) == 0 {
		return nil, fmt.Errorf("failed to marshal: %w", err)
	}
	return out, nil // error is safe to ignore as long as we have the payload bytes
}

var _ jose.Signer = payloadGrabber(nil)

type payloadGrabber func(payload []byte)

func (p payloadGrabber) Sign(payload []byte) (*jose.JSONWebSignature, error) {
	p(payload)
	return nil, jose.ErrUnprotectedNonce // return some error to stop after we have the payload
}

func (p payloadGrabber) Options() jose.SignerOptions { return jose.SignerOptions{} }
```

Reply:
I had the same thought at first (and agree on using the same logic if possible), but I thought the jose.Signer got the full payload including the header, which we were leaving to the external signer to generate

Reply:
Done